src/NuGet.Clients/VsConsole/PowerShellHost/PowerShellHost.cs

```diff
@@ -362,16 +381,30 @@ private void UpdateWorkingDirectory()
                     return;
                 }
 
-                // check if we have already run init scripts for this solution, then simply return without executing init scripts
-                if (!string.IsNullOrEmpty(_currentSolutionDirectory) &&
-                    _currentSolutionDirectory.Equals(_solutionManager.SolutionDirectory, StringComparison.OrdinalIgnoreCase))
+                // This value is set outside of the lock, so cache the value during the course of the execution.
+                var latestSolutionSpecHashReceived = _latestSolutionSpecHashReceived;
+
+                // Only no-op using solution spec hash if we have received a restore event.
+                if (latestSolutionSpecHashReceived == null)
                 {
-                    return;
+                    // check if we have already run init scripts for this solution, then simply return without executing init scripts
+                    if (!string.IsNullOrEmpty(_currentSolutionDirectory) &&
+                        _currentSolutionDirectory.Equals(_solutionManager.SolutionDirectory, StringComparison.OrdinalIgnoreCase))
+                    {
+                        return;
+                    }
+                    else
+                    {
+                        // save current solution directory so that we don't execute init scripts for the same solution again.
+                        _currentSolutionDirectory = _solutionManager.SolutionDirectory;
+                    }
                 }
-                else
+                else if (latestSolutionSpecHashReceived == _latestSolutionSpecHashExecuted)
```

Comment:
Here also you need to check for `_currentSolutionDirectory` change since there could be a scenario that restore happend on solution1 which updated `_latestSolutionSpecHashExecuted` but then user changed the solution and tried to install a package on packages.config based project which will make it bypass executing init.ps1.